File: week09/15/1_Mon_ValidPath.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<vector<int>> neighbors(n);
        for (const auto& edge : edges) {
            neighbors[edge[0]].push_back(edge[1]);
            neighbors[edge[1]].push_back(edge[0]);
        }

        vector<bool> visit(n, false);
        dfs(source, neighbors, visit);
        return visit[destination];
    }
    void dfs(int node, const vector<vector<int>>& neighbors, vector<bool>& visit) {
        visit[node] = true;
        for (int neighbor : neighbors[node]) {
            if (!visit[neighbor]) {
                dfs(neighbor, neighbors, visit);
            }
        }
    }
};

File: week09/15/1_Mon_ValidPath.py
from typing import List

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        neighbors = [[] for _ in range(n)]
        for edge in edges:
            neighbors[edge[0]].append(edge[1])
            neighbors[edge[1]].append(edge[0])
            
        visited = [False] * n
        def help(node):
            visited[node] = True
            for neighbor in neighbors[node]:
                if not visited[neighbor]:                    
                    help(neighbor)
        help(source)
        return visited[destination]